app/services/notify_service.rb

```diff
@@ -16,13 +16,86 @@ class NotifyService < BaseService
   ).freeze
 
-  class DismissCondition
+  class BaseCondition
+    NEW_ACCOUNT_THRESHOLD = 30.days.freeze
+
+    NEW_FOLLOWER_THRESHOLD = 3.days.freeze
+
+    NON_FILTERABLE_TYPES = %i(
+      admin.sign_up
+      admin.report
+      poll
+      update
+      account_warning
+    ).freeze
+
     def initialize(notification)
       @recipient = notification.account
       @sender = notification.from_account
       @notification = notification
+      @policy = NotificationPolicy.find_or_initialize_by(account: @recipient)
+    end
+
+    private
+
+    def not_following?
+      !@recipient.following?(@sender)
     end
 
+    def not_follower?
+      follow = Follow.find_by(account: @sender, target_account: @recipient)
+      follow.nil? || follow.created_at > NEW_FOLLOWER_THRESHOLD.ago
+    end
+
+    def new_account?
+      @sender.created_at > NEW_ACCOUNT_THRESHOLD.ago
+    end
+
+    def override_for_sender?
+      NotificationPermission.exists?(account: @recipient, from_account: @sender)
+    end
+
+    def from_limited?
+      @sender.silenced? && not_following?
+    end
+
+    def private_mention_not_in_response?
+      @notification.type == :mention && @notification.target_status.direct_visibility? && !response_to_recipient?
+    end
+
+    def response_to_recipient?
+      return false if @notification.target_status.in_reply_to_id.nil?
+
+      statuses_that_mention_sender.positive?
+    end
+
+    def statuses_that_mention_sender
+      # This queries private mentions from the recipient to the sender up in the thread.
+      # This allows up to 100 messages that do not match in the thread, allowing conversations
+      # involving multiple people.
+      Status.count_by_sql([<<-SQL.squish, id: @notification.target_status.in_reply_to_id, recipient_id: @recipient.id, sender_id: @sender.id, depth_limit: 100])
+        WITH RECURSIVE ancestors(id, in_reply_to_id, mention_id, path, depth) AS (
+            SELECT s.id, s.in_reply_to_id, m.id, ARRAY[s.id], 0
+            FROM statuses s
+            LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id
+            WHERE s.id = :id
+          UNION ALL
+            SELECT s.id, s.in_reply_to_id, m.id, ancestors.path || s.id, ancestors.depth + 1
+            FROM ancestors
+            JOIN statuses s ON s.id = ancestors.in_reply_to_id
+            /* early exit if we already have a mention matching our requirements */
+            LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id AND s.account_id = :recipient_id
+            WHERE ancestors.mention_id IS NULL AND NOT s.id = ANY(path) AND ancestors.depth < :depth_limit
+        )
+        SELECT COUNT(*)
+        FROM ancestors
+        JOIN statuses s ON s.id = ancestors.id
+        WHERE ancestors.mention_id IS NOT NULL AND s.account_id = :recipient_id AND s.visibility = 3
+      SQL
+    end
+  end
+
+  class DismissCondition < BaseCondition
```

Comment:
It might make sense to rename this to `DropCondition` since the option on the policy now uses the value "drop".

Reply:
Renamed to `DropCondition`, and renamed `DismissCondition#dismiss?` to `DropCondition#drop?`. Though I wonder if we really want those to be two distinct classes and objects, rather than, say, a single `FilterCondition` class with a method `drop?` and a method `filter?`